src/Microsoft.Diagnostics.Monitoring/DiagnosticsMonitor.cs

```diff
@@ -19,359 +19,126 @@ namespace Microsoft.Diagnostics.Monitoring
 
-            for (int i = 0; i < 10; ++i)
+            lock (_lock)
             {
-                if (DiagnosticsClient.GetPublishedProcesses().Contains(processId))
+                if (_disposed)
                 {
-                    hasEventPipe = true;
-                    break;
+                    throw new ObjectDisposedException(nameof(DiagnosticsMonitor));
                 }
 
-                cancellationToken.ThrowIfCancellationRequested();
-
-                await Task.Delay(500);
-            }
-
-            if (!hasEventPipe)
-            {
-                _logger.LogInformation("Process id {PID}, does not support event pipe", processId);
-                return;
-            }
-
-            _logger.LogInformation("Listening for event pipe events for {ServiceName} on process id {PID}", _dimValues[1], processId);
+                if (_currentTask != null)
+                {
+                    throw new InvalidOperationException("Only one stream processing is allowed");
+                }
 
-            while (!cancellationToken.IsCancellationRequested)
-            {
                 EventPipeSession session = null;
                 var client = new DiagnosticsClient(processId);
 
                 try
                 {
                     session = client.StartEventPipeSession(_sourceConfig.GetProviders());
                 }
-                catch (EndOfStreamException)
+                catch (EndOfStreamException e)
                 {
-                    break;
+                    throw new InvalidOperationException("End of stream", e);
                 }
-                catch (Exception ex)
+                catch (Exception ex) when (!(ex is OperationCanceledException))
                 {
-                    if (!cancellationToken.IsCancellationRequested)
-                    {
-                        _logger.LogDebug(0, ex, "Failed to start the event pipe session");
-                    }
-
-                    // We can't even start the session, wait until the process boots up again to start another metrics thread
-                    break;
+                    throw new InvalidOperationException("Failed to start the event pipe session", ex);
                 }
 
-                void StopSession()
+                CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(_disposeSource.Token, cancellationToken);
+
+                _currentTask = Task.Run( async () =>
                 {
                     try
                     {
-                        session.Stop();
-                    }
-                    catch (EndOfStreamException)
-                    {
-                        // If the app we're monitoring exits abruptly, this may throw in which case we just swallow the exception and exit gracefully.
+                        int durationMilliseconds = (durationSeconds != -1) ? durationSeconds * 1000 : durationSeconds;
```

Comment:
I think you can simplify this by changing all of the duration parameters from int to TimeSpan. It allows the caller to provide whatever time interval it wants, avoids having to specify in the parameter name the unit of time, allows for more granular or extremely large time intervals, and makes it easy to specify infinite timeout/duration (using Timeout.InfiniteTimespan) without having to explicitly handle conversion in this code.

Reply:
While I agree from an Api perspective, this may introduce more difficulties at the REST layer. I will defer this for another PR.

Reply:
I started looking at this in some changes that I was making, so if you don't get to it, I can take care of it.